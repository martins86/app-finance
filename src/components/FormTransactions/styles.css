.form-transactions {
  max-width: 1120px;
  width: 92%;
  margin: 0 auto;
  display: flex;
  justify-content: space-around;
  background-color: #ffffff;
  border-radius: 5px;
  padding: 15px;
  box-shadow: 3px 2px 4px 1px #00000054;
  gap: 10px;
}

.form-field {
  display: flex;
  flex-direction: column;
  width: 20%;
}

.form-radio-group {
  display: flex;
  align-items: center;
  width: 220px;
  margin-top: 20px;
}

.input-radio {
  margin-left: 20px;
  margin-right: 5px;
  accent-color: #d61b1b;
  margin-top: 0;
}

@media (max-width: 750px) {
  .form-transactions {
    display: grid;
  }
}
