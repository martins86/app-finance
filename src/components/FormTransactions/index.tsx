import { useState } from 'react';

import Button from './../../components/Button';
import InputType from './../../components/InputType';

import './styles.css';

const FormTransactions = ({ handleAdd }: any) => {
  const [desc, setDesc] = useState();
  const [amount, setAmount] = useState();
  const [isExpense, setExpense] = useState(false);
  const [symbolsArr] = useState(['e', 'E', '+', '-']);

  const generateID = () => Math.round(Math.random() * 10000);

  const handleSave = () => {
    if (!desc || !amount) {
      alert('Alerta:\nInforme a descrição e o valor!');
      return;
    } else if (amount < 1) {
      alert('Alerta:\nO valor tem que ser positivo!');
      return;
    }

    const transaction = {
      id: generateID(),
      desc: desc,
      amount: amount,
      expense: isExpense,
    };

    handleAdd(transaction);

    setDesc(undefined);
    setAmount(undefined);
  };

  return (
    <div className="form-transactions">
      <div className="form-fieldset">
        <label htmlFor="desc">Descrição</label>
        <InputType
          type="text"
          id="desc"
          dataTestId="desc"
          className="input-primary"
          placeholder="Digite a descrição"
          value={desc || ''}
          onChange={(e: any) => setDesc(e.target.value)}
        />
      </div>
      <div className="form-fieldset">
        <label htmlFor="amount">Valor</label>
        <InputType
          type="number"
          id="amount"
          dataTestId="amount"
          className="input-primary"
          placeholder="Digite o valor"
          value={amount || ''}
          min="1"
          onKeyDown={(e: any) =>
            symbolsArr.includes(e.key) && e.preventDefault()
          }
          onChange={(e: any) => setAmount(e.target.value)}
        />
      </div>
      <div className="form-fieldset">
        <div className="form-radio-group">
          <div className="fieldset-item">
            <InputType
              type="radio"
              id="rIncome"
              dataTestId="radio-income"
              nameField="group1"
              className="input-radio"
              defaultChecked
              onChange={() => setExpense(!isExpense)}
            />
            <label htmlFor="rIncome" title="Checar como entrada">
              Entrada
            </label>
          </div>

          <div className="fieldset-item">
            <InputType
              type="radio"
              id="rExpenses"
              dataTestId="radio-expenses"
              nameField="group1"
              className="input-radio"
              onChange={() => setExpense(!isExpense)}
            />
            <label htmlFor="rExpenses" title="Checar como saída">
              Saída
            </label>
          </div>
        </div>
      </div>
      <div className="form-fieldset">
        <Button
          className="btn btn--primary"
          text="Adicionar"
          onClick={handleSave}
        />
      </div>
    </div>
  );
};

export default FormTransactions;
