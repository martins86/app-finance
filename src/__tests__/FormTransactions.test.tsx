import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import FormTransactions from '../components/FormTransactions';

describe('Test <FormTransactions />', () => {
  test('Should render FormTransactions with three inputs', () => {
    // Arrange
    render(
      <Router>
        <FormTransactions />
      </Router>
    );

    // Act
    const elInputText = screen.getByPlaceholderText(/Digite a descrição/i);
    const elInputNumber = screen.getByPlaceholderText(/Digite o valor/i);
    const elInputRadioOne = screen.getByTestId('radio-income');
    const elInputRadioTwo = screen.getByTestId('radio-expenses');

    // Assert
    expect(elInputText).toBeInTheDocument();
    expect(elInputNumber).toBeInTheDocument();
    expect(elInputRadioOne).toBeInTheDocument();
    expect(elInputRadioTwo).toBeInTheDocument();
  });

  test('Should render FormTransactions with one button "Adicionar"', () => {
    // Arrange
    const fn = jest.fn();
    render(
      <Router>
        <FormTransactions handleAdd={fn} />
      </Router>
    );

    // Act
    const elButton = screen.getByRole('button', { name: /Adicionar/i });

    // Assert
    expect(elButton).toBeInTheDocument();
  });

  test('Should match snapshot', () => {
    // Arrange
    const fn = jest.fn();
    const { container } = render(
      <Router>
        <FormTransactions handleAdd={fn} />
      </Router>
    );

    // Act & Assert
    expect(container).toMatchSnapshot();
  });
});
