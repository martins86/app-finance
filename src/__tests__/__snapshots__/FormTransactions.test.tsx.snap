// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <FormTransactions /> Should match snapshot 1`] = `
<div>
  <div
    class="form-transactions"
  >
    <div
      class="form-fieldset"
    >
      <label
        for="desc"
      >
        Descrição
      </label>
      <input
        class="input-primary"
        data-testid="desc"
        id="desc"
        placeholder="Digite a descrição"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-fieldset"
    >
      <label
        for="amount"
      >
        Valor
      </label>
      <input
        class="input-primary"
        data-testid="amount"
        id="amount"
        min="1"
        placeholder="Digite o valor"
        type="number"
        value=""
      />
    </div>
    <div
      class="form-fieldset"
    >
      <div
        class="form-radio-group"
      >
        <div
          class="fieldset-item"
        >
          <input
            checked=""
            class="input-radio"
            data-testid="radio-income"
            id="rIncome"
            name="group1"
            type="radio"
            value=""
          />
          <label
            for="rIncome"
            title="Checar como entrada"
          >
            Entrada
          </label>
        </div>
        <div
          class="fieldset-item"
        >
          <input
            class="input-radio"
            data-testid="radio-expenses"
            id="rExpenses"
            name="group1"
            type="radio"
            value=""
          />
          <label
            for="rExpenses"
            title="Checar como saída"
          >
            Saída
          </label>
        </div>
      </div>
    </div>
    <div
      class="form-fieldset"
    >
      <button
        class="btn btn--primary"
        data-testid="button"
        title="Adicionar"
        type="button"
      >
        Adicionar
      </button>
    </div>
  </div>
</div>
`;
